File: 2021/js/Day-04/Day-04.js
const fs = require("fs");

inputData = fs.readFileSync("input.txt", "utf8");

inputData = inputData.split(/\r?\n/).map((v) => v.trim());

// parse input data
let numbers = [],
  currentBoard = [],
  boards = [];

for (const [i, line] of inputData.entries()) {
  // console.log(i, inputData.length);

  //get the numbers
  if (i === 0) {
    for (const n of line.split(",")) {
      numbers.push(Number(n));
    }

    continue;
  }

  currentBoard.push(
    line
      .split(" ")
      .filter((v) => v) //remove empty entries
      .map((v) => Number(v))
  );

  //empty, or last, line pushed currentBoard to boards
  if (!line || i === inputData.length - 1) {
    if (!currentBoard.length) continue;

    if (currentBoard.length > 1) boards.push(currentBoard.slice(0, 5));
    currentBoard = [];
    continue;
  }
}

const transpose = (matrix) => {
  return matrix[0].map((col, i) => matrix.map((row) => row[i]));
};

const checkForWin = (board, calledNumbers) => {
  // console.log(board);
  for (const row of [...board, ...transpose(board)]) {
    if (row.every((v) => calledNumbers.includes(v))) {
      return true;
    }
  }

  return false;
};

const sumOfBoard = (board, calledNumbers) => {
  return board
    .flat()
    .reduce((sum, v) => (!calledNumbers.includes(v) ? sum + v : sum), 0);
};

const Solution01 = (numbers, boards) => {
  const findFirstWinner = (numbers, boards) => {
    let calledNumbers = [];

    //game loop
    for (const num of numbers) {
      calledNumbers.push(num);

      //check for winners
      for ([i, board] of boards.entries()) {
        const isWinner = checkForWin(board, calledNumbers);

        if (isWinner) {
          return [i, calledNumbers.at(-1), calledNumbers];
        }
      }
    }
  };

  //start the bingo
  const [winningBoardIndex, lastCalledNum, calledNumbers] = findFirstWinner(
    numbers,
    boards
  );

  const sumOfWinningBoard = sumOfBoard(
    boards[winningBoardIndex],
    calledNumbers
  );

  return sumOfWinningBoard * lastCalledNum;
};

const Solution02 = (numbers, boards) => {
  let calledNumbers = [];
  let winningBoards = [];
  let calledNumbersAtLastWin = [];
  let boardIndexAtLastWin;

  //game loop
  for (const num of numbers) {
    calledNumbers.push(num);

    //check for winners
    for ([i, board] of boards.entries()) {
      if (winningBoards.includes(i)) continue;

      const isWinner = checkForWin(board, calledNumbers);

      if (isWinner) {
        winningBoards.push(i);
        boardIndexAtLastWin = i;
        calledNumbersAtLastWin = [...calledNumbers];
      }
    }
  }

  const sumOfLastWinningBoard = sumOfBoard(
    boards[boardIndexAtLastWin],
    calledNumbersAtLastWin
  );

  return sumOfLastWinningBoard * calledNumbersAtLastWin.at(-1);
};

console.log(`Solution 01: ${Solution01(numbers, boards)}`);
console.log(`Solution 02: ${Solution02(numbers, boards)}`);

File: 2021/python/Day-04/Day_04.py
import time
import copy



with open("input.txt","r") as file:
    boards = []
    current_board = []

    for i, line in enumerate([line.strip() for line in file]):
        if i == 0:  #first row
            numbers = [int(number) for number in line.split(",")]

        else:
            if not line:
                continue

            current_board.append([{'number':int(number), 'hit':False} for number in line.split()])

            if len(current_board) == 5:
                boards.append(current_board)
                current_board = []




def solution01(input_boards, numbers):
    boards = copy.deepcopy(input_boards)

    def bingo():
        for called_number_i, called_number in enumerate(numbers):
            #Mark hits
            for board_i, board in enumerate(boards):
                for row_i, row in enumerate(board):
                    for num_i, num in enumerate(row):
                        if num['number'] == called_number:
                            boards[board_i][row_i][num_i]['hit'] = True

            #Check for bingo
            if called_number_i > 3:   #skip check for the first 4 numbers
                for board_i, board in enumerate(boards):
                    if any(all(num['hit'] for num in row) for row in board) or any(all(num['hit'] for num in col) for col in zip(*board)):
                        return board_i, called_number
                     

    board_with_bingo, last_number_called = bingo()
    non_hit_numbers = []
    for row in boards[board_with_bingo]:
        for number in row:
            if not number['hit']:
                non_hit_numbers.append(number['number'])

    return sum(non_hit_numbers) * last_number_called


def solution02(input_boards, numbers):
    boards = copy.deepcopy(input_boards)

    def bingo():
        boards_with_bingo = []
        check_counter = 0
        for called_number_i, called_number in enumerate(numbers):
            #Mark hits
            for board_i, board in enumerate(boards):
                if board_i in [board[0] for board in boards_with_bingo]:
                    continue

                for row_i, row in enumerate(board):
                    for num_i, num in enumerate(row):
                        if num['number'] == called_number:
                            boards[board_i][row_i][num_i]['hit'] = True
        
            #Check for bingo
            if called_number_i > 3:   #skip check for the first 4 numbers
                for board_i, board in enumerate(boards):
                    if board_i in [board[0] for board in boards_with_bingo]:
                        continue

                    if any(all(num['hit'] for num in row) for row in board) or any(all(num['hit'] for num in col) for col in zip(*board)):
                        boards_with_bingo.append((board_i, called_number))
        
        return boards_with_bingo[-1] 


    last_board_with_bingo, number_called_at_last_bingo = bingo()
    non_hit_numbers = []
    for row in boards[last_board_with_bingo]:
        for number in row:
            if not number['hit']:
                non_hit_numbers.append(number['number'])

    return sum(non_hit_numbers) * number_called_at_last_bingo

solution01_start_time = time.time()
solution01 = solution01(boards, numbers)
solution01_total_time = time.time() - solution01_start_time
print(f"Solution 01: {solution01}, in {solution01_total_time}")

solution02_start_time = time.time()
solution02 = solution02(boards, numbers)
solution02_total_time = time.time() - solution02_start_time
print(f"Solution 02: {solution02}, in {solution02_total_time}")